pkg/scheduling/volumelimits.go

```diff
@@ -0,0 +1,94 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// VolumeLimits tracks volume limits on a per node basis.  The number of volumes that can be mounted varies by instance
+// type. We need to be aware and track the mounted volume usage to inform our awareness of which pods can schedule to
+// which nodes.
+type VolumeLimits struct {
+	ctx        context.Context
+	volumes    sets.String
+	podVolumes map[types.NamespacedName]sets.String
+}
+
+func NewVolumeLimits(ctx context.Context) *VolumeLimits {
+	return &VolumeLimits{
+		ctx:        ctx,
+		volumes:    sets.NewString(),
+		podVolumes: map[types.NamespacedName]sets.String{},
+	}
+}
+
+func (v *VolumeLimits) Add(pod *v1.Pod) {
+	podVolumes, err := v.validate(pod)
+	if err != nil {
+		logging.FromContext(v.ctx).Errorf("inconsistent state registering volume li mit, %s", err)
```

Comment:
This seems like an antipattern to have all of these ERROR messages complaining about invariants. What's a customer to do about it? Can we structure the code such that this simply can't happen?

Reply:
This particular one is gone, there is still one or two remaining in the host port usage.  It indicates that an assumption we've made has not held true, ideally customers don't see it and if they do we know to re-look at the assumption vs troubleshooting a hard to find logic error.

Reply:
Can we make the message something like "Invariant violated, please cut an issue"